api/src/users/users.service.ts

```diff
@@ -1,100 +1,241 @@
+import { User } from 'src/types/schema';
+import { Selectable } from 'kysely';
 
 @Injectable()
 export class UsersService {
-  async createUser(createUsersDto: CreateUsersDto): Promise<string> {
+  /**
+   * Create a user in the database using createUserDto values
+   * @param createUsersDto 
+   * @returns AppUser
+   * @throws InternalServerError
+   * @throws UnprocessableEntity
+   */
+  async createUser(createUsersDto: CreateUsersDto): Promise<AppUser> {
+    if (createUsersDto.username == "" || !createUsersDto.username) {
+      console.log("Tried to register without a username");
+      throw new UnprocessableEntityException("Empty username");
+    }
     try {
       const hashedPassword = await bcrypt.hash(createUsersDto.password, 10);
-      if (createUsersDto.username == "")
-        throw "Empty username";
-      await db
+       const result = await db
       .insertInto('user')
       .values({
         username: createUsersDto.username,
         password: hashedPassword,
         firstname: createUsersDto.firstname,
         lastname: createUsersDto.lastname
       })
-      .executeTakeFirstOrThrow();
-      return 'success';
+      .executeTakeFirst();
+      if (!result.numInsertedOrUpdatedRows || result.numInsertedOrUpdatedRows <= 0n)
+        throw new UnprocessableEntityException("Username already taken");
+    } catch (error) {
+      console.log(error);
+      if (error instanceof UnprocessableEntityException)
+        throw error;
+      throw new InternalServerErrorException();
+    }
+    let user: AppUser | undefined;
+    try {
+      user = await db
+      .selectFrom('user')
+      .selectAll()
+      .where('username', '=', createUsersDto.username)
+      .executeTakeFirstOrThrow()
     } catch (error) {
-      return 'error';
+      console.log(error);
+      throw new InternalServerErrorException();
     }
+    return user;
   }
 
-  async getUserById(userId: number): Promise<AppUser | undefined> {
+  /**
+   * Create a user in the db for oauth, we generate a password for the user,
+   * but we don't communicate to them, since they will use oauth to connect.
+   * They will be able to change the password if they want to connect without
+   * oauth though.
+   * @param intraUser 
+   * @returns True if we correctly create a new user, false otherwise
+   * @throws InternalServerError if the db fail
+   */
+  async createOauthUser(intraUser: userFromIntra): Promise<boolean> {
+    try {
+      const hashedPassword = await bcrypt.hash(randomBytes(32), 10);
+      const result = await db
+      .insertInto('user')
+      .values({
+        email: intraUser.email,
+        username: intraUser.username,
+        avatarUrl: intraUser.avatarUrl,
+        firstname: intraUser.firstname,
+        lastname: intraUser.lastname,
+        password: hashedPassword
+      })
+      .executeTakeFirst();
+      if (result.numInsertedOrUpdatedRows === 0n)
+        return false;
+      return true;
```

Comment:
Wouldn't it be usefull to return the created user here?

Reply:
No, we already have all the information we need for the user created through API 42.